### DISCUSSION TITLE
Ws close handler not called when server is gracefully shutdown

### DISCUSSION THREAD

**Discussion by dandoen** [uNetworking/uWebSockets.js#299] [Q&A] [answered]
Provided you offload some state as to who is connected, and you shutdown the server gracefully on SIGINT (i.e. scaling down), what's the approach one needs to take to invoke the close handler so the external state can be updated with the latest status for the now disconnected clients.

I am simply doing this from the graceful shutdown example and not seeing the close handler get invoked, like it does when the client disconnects.

```
setTimeout(() => {
  console.log('Shutting down now');
  uWS.us_listen_socket_close(listenSocket);
  listenSocket = null;
}, 1000);
```

I can manually keep track of connected clients and do that in a manual drain call, but wondering if there's a better way/there's something I'm missing.

**Comment by dandoen** [comment]
@alexhultman can you shed some light as to if this is a correct read of what's going on in this case:

When WebSocket.end gets called, it sets `webSocketData->isShuttingDown` to `true`.
https://github.com/uNetworking/uWebSockets/blob/3a856b610bdaae312969cb93ac9fba823913b08e/src/WebSocket.h#L127

Then when `us_socket_context_on_close` gets invoked, this check specifically prevents from the `closeHandler` to get invoked. https://github.com/uNetworking/uWebSockets/blob/3a856b610bdaae312969cb93ac9fba823913b08e/src/WebSocketContext.h#L243

Wondering what the rationale is behind this?

**Comment by ghost** [comment, marked as answer]
This is by design. You're closing the listen socket. This has nothing to do with the websockets, they remain open.

If you want to close the websockets then you call close on them

**Comment by dandoen** [comment]
@alexhultman The server is closed and the connections are closed so the websockets do not "remain open" so unclear what you are talking about?

**Comment by ghost** [comment]
If your websockets **are** closed then of course they won't invoke its close handler, as they **already** are closed. So I fail to follow the logic in your report. I can recommend reading the user manual, it is very clear on this topic.
